core/src/main/scala/org/apache/spark/deploy/SparkS3Util.scala

```diff
@@ -0,0 +1,336 @@
+      newClient
+    }
+  }
+
+  /**
+   * Helper function to extract S3 object key name from the given path.
+   */
+  private def keyFromPath(path: Path): String = {
+    Preconditions.checkArgument(path.isAbsolute, "Path is not absolute: %s", path)
+    var key: String = Strings.nullToEmpty(path.toUri.getPath)
+    if (key.startsWith("/")) {
+      key = key.substring(1)
+    }
+    if (key.endsWith("/")) {
+      key = key.substring(0, key.length - 1)
+    }
+    key
+  }
+
+  /**
+   * Helper function to convert S3ObjectSummary into FileStatus.
+   */
+  private def statusFromObjects(
+      bucket: String,
+      objects: util.List[S3ObjectSummary]): Iterator[FileStatus] = {
+    val blockSize: Long = getS3BlockSize()
+    val list: ArrayBuffer[FileStatus] = ArrayBuffer()
+    for (obj: S3ObjectSummary <- objects.asScala) {
+      if (!obj.getKey.endsWith("/")) {
+        val path = new Path(bucket + "/" + obj.getKey)
+        if (hiddenFileFilter.accept(path)) {
+          val status: FileStatus = new FileStatus(
+            obj.getSize,
+            false,
+            1,
+            blockSize,
+            obj.getLastModified.getTime,
+            path)
+          list += status
+        }
+      }
+    }
+    list.iterator
+  }
+
+  /**
+   * For the given path, list S3 objects and return an iterator of returned FileStatuses.
+   */
+  @throws(classOf[AmazonClientException])
+  @throws(classOf[AmazonServiceException])
+  private def listPrefix(s3: AmazonS3Client, path: Path): Iterator[FileStatus] = {
+    val uri: URI = path.toUri
+    val key: String = keyFromPath(path)
+    val request: ListObjectsRequest = new ListObjectsRequest()
+      .withBucketName(uri.getAuthority)
+      .withPrefix(key)
+
+    val listings = new AbstractSequentialIterator[ObjectListing](s3.listObjects(request)) {
+      protected def computeNext(previous: ObjectListing): ObjectListing = {
+        if (!previous.isTruncated) {
+          return null
+        }
+        s3.listNextBatchOfObjects(previous)
+      }
+    }.asScala
+
+    val bucket: String = uri.getScheme + "://" + uri.getAuthority
+    listings
+      .map(listing => statusFromObjects(bucket, listing.getObjectSummaries))
+      .reduceLeft(_ ++ _)
+  }
+
+  /**
+   * For the given list of paths, sequentially list S3 objects per path and combine returned
+   * FileStatuses into a single array.
+   */
+  private def listStatus(s3: AmazonS3Client, paths: List[Path]): Array[FileStatus] = {
+    val list: ArrayBuffer[FileStatus] = ArrayBuffer()
+    paths.foreach { path =>
+      val iterator = listPrefix(s3, path)
+      while (iterator.hasNext) {
+        list += iterator.next
+      }
+    }
+    list.toArray
+  }
+
+  /**
+   * Find S3 block size from Hadoop conf. Try both s3 and s3n names.
+   */
+  private def getS3BlockSize(): Long = {
+    val value = Option(conf.get("fs.s3.block.size"))
+      .getOrElse(conf.get("fs.s3n.block.size", "67108864"))
+    value.toLong
```

Comment:
@steveloughran Thanks for the comment. I agree. What size of a min blocksize do you think is reasonable for such a check? 10mb?